File: app/src/test/java/com/renanferrari/testeandroidv2/application/ui/statements/StatementsViewModelTest.java
package com.renanferrari.testeandroidv2.application.ui.statements;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.Observer;
import com.renanferrari.testeandroidv2.application.common.providers.SchedulerProvider;
import com.renanferrari.testeandroidv2.application.common.utils.AccountFormatter;
import com.renanferrari.testeandroidv2.application.common.utils.DateFormatter;
import com.renanferrari.testeandroidv2.application.common.utils.MoneyFormatter;
import com.renanferrari.testeandroidv2.domain.interactors.GetStatements;
import com.renanferrari.testeandroidv2.domain.interactors.GetUser;
import com.renanferrari.testeandroidv2.domain.interactors.Logout;
import io.reactivex.Maybe;
import io.reactivex.Single;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static com.google.common.truth.Truth.assertThat;
import static com.renanferrari.testeandroidv2.common.Constants.VALID_STATEMENT_LIST;
import static com.renanferrari.testeandroidv2.common.Constants.VALID_USER;
import static com.renanferrari.testeandroidv2.common.Constants.VALID_USER_ACCOUNT;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class) public class StatementsViewModelTest {

  @Mock private GetStatements getStatements;
  @Mock private GetUser getUser;
  @Mock private Logout logout;

  @Mock private Observer<UserState> userStateObserver;
  @Mock private Observer<StatementsState> statementsStateObserver;

  @Captor private ArgumentCaptor<UserState> userStateArgumentCaptor;
  @Captor private ArgumentCaptor<StatementsState> statementsStateArgumentCaptor;

  @Rule public TestRule instantTaskExecutorRule = new InstantTaskExecutorRule();

  private MoneyFormatter moneyFormatter;
  private DateFormatter dateFormatter;
  private StatementsViewModel statementsViewModel;

  @Before public void setUp() {
    final Locale locale = new Locale("pt", "BR");
    moneyFormatter = new MoneyFormatter(locale);
    dateFormatter = new DateFormatter(locale);
    statementsViewModel = new StatementsViewModel(getStatements, getUser, logout, moneyFormatter,
        dateFormatter, SchedulerProvider.forTesting());
  }

  @Test public void onUserRequested_withLoggedInUser_shouldUpdateState() {
    when(getUser.execute()).thenReturn(Maybe.just(VALID_USER));

    statementsViewModel.getObservableUserState().observeForever(userStateObserver);
    statementsViewModel.onUserRequested();

    final UserState expectedFinalState = UserState.builder()
        .name(VALID_USER.getName())
        .account(AccountFormatter.format(VALID_USER.getBankAccount(), VALID_USER.getAgency()))
        .balance(moneyFormatter.format(VALID_USER.getBalance()))
        .build();

    verify(userStateObserver, times(1)).onChanged(userStateArgumentCaptor.capture());

    assertThat(userStateArgumentCaptor.getValue()).isEqualTo(expectedFinalState);
  }

  @Test public void onUserRequested_withNoUser_shouldNotUpdateState() {
    when(getUser.execute()).thenReturn(Maybe.empty());

    statementsViewModel.getObservableUserState().observeForever(userStateObserver);
    statementsViewModel.onUserRequested();

    verifyZeroInteractions(userStateObserver);
  }

  @Test public void onStatementsRequested() {
    when(getStatements.execute()).thenReturn(Single.just(VALID_STATEMENT_LIST));

    statementsViewModel.getObservableStatementsState().observeForever(statementsStateObserver);
    statementsViewModel.onStatementsRequested();

    final StatementsState expectedInitialState = StatementsState.createDefault();
    final StatementsState expectedLoadingState = expectedInitialState.withLoading(true);
    final StatementsState expectedFinalState = StatementsState.create(Arrays.asList(
        StatementItem.create(VALID_STATEMENT_LIST.get(0), moneyFormatter, dateFormatter),
        StatementItem.create(VALID_STATEMENT_LIST.get(1), moneyFormatter, dateFormatter),
        StatementItem.create(VALID_STATEMENT_LIST.get(2), moneyFormatter, dateFormatter)));

    verify(statementsStateObserver, times(3)).onChanged(statementsStateArgumentCaptor.capture());

    final List<StatementsState> allStates = statementsStateArgumentCaptor.getAllValues();

    assertThat(allStates.get(0)).named("initial state").isEqualTo(expectedInitialState);
    assertThat(allStates.get(1)).named("loading state").isEqualTo(expectedLoadingState);
    assertThat(allStates.get(2)).named("final state").isEqualTo(expectedFinalState);
  }
}

File: presentation/src/test/java/br/com/silas/testeandroidv2/viewModels/StatementsViewModelTest.kt
package br.com.silas.testeandroidv2.br.com.silas.testeandroidv2.viewModels

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import br.com.silas.domain.statements.StatementsInteractor
import br.com.silas.testeandroidv2.any
import br.com.silas.testeandroidv2.br.com.silas.testeandroidv2.mocks.ErrorResponseMock
import br.com.silas.testeandroidv2.br.com.silas.testeandroidv2.mocks.StatementsMock
import br.com.silas.testeandroidv2.ui.statements.StatementsViewModel
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Single
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class StatementsViewModelTest {

    @Mock
    lateinit var statementsInteractor: StatementsInteractor

    lateinit var statementsViewModel: StatementsViewModel

    @get:Rule
    var rule: TestRule = InstantTaskExecutorRule()

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        statementsViewModel = StatementsViewModel(statementsInteractor)
    }

    @Test
    fun `When fetch statements is successful should be return a list of statements`() {
        val statementsList = StatementsMock.getListStatements()
        val statementsError = ErrorResponseMock.getErrorResponseIsEmpty()
        val pairResponse = Pair(statementsError, statementsList)

        `when`(statementsInteractor.execute(any())).thenReturn(Single.just(pairResponse))
        statementsViewModel.loadStatements(152)

        assertThat(statementsViewModel.result.value, `is`(pairResponse.second))
    }

    @Test
    fun `When fetch statements is unsuccessful should be return an error response`() {
        val statementsList = StatementsMock.getListStatements()
        val statementsError = ErrorResponseMock.getErrorResponseIsNotEmpty()
        val pairResponse = Pair(statementsError, statementsList)

        `when`(statementsInteractor.execute(any())).thenReturn(Single.just(pairResponse))
        statementsViewModel.loadStatements(152)

        assertThat(statementsViewModel.errorStatements.value, `is`(pairResponse.first))
    }
}